### PR TITLE
 Refactored 'tags' validation. Moved all tags tests into single class.  Added new test case to validate datasource tags.

### PR DESCRIPTION
**Pull request by hhovsepy**
Changed Select to AngularSelect for tags input selects in mixins.
Added new test case to validate datasource tags.
IMPORTANT:  Changed Providers removing menu item text. It is changed to "Remove {} Providers" @psav @mfalesni 
Fixed failing topology test.

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 5135f221b94ad34588a70c71598ea7ece3eca00d:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 6907cff5b2898fdefb01467d7fe7309e36dcf340:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by hhovsepy** [comment]
In fact tests are skipped, because of "ERROR:cfme:Could not do is_refreshed at /cfme_tests_te/utils/varmeth.py:74 in time"

**Comment by dajoRH** [comment]
Lint report for commit 10a9b92d7fc3a127bad28e73f16ef7baf2e22678:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 59b3206829f289e117f913d0aa3467f22cabcc52:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit a6be0e23ed19c1d4528909823edfca04caca097a:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 0320ba30ee82c125ae640c06e2ed73b3be1c64ff:

Everything seems all right :smile: .
*CFME QE Bot*

File: cfme/tests/middleware/test_middleware_datasource.py [hunk 1]
```diff
@@ -122,3 +123,22 @@ def _get_datasources_set(datasources):
...
+    tags = [
+        Tag(category=Category(display_name='Environment', single_value=True), display_name='Test'),
+        Tag(category=Category(display_name='Location', single_value=True), display_name='New York'),
+        Tag(category=Category(display_name='Workload', single_value=False),
+            display_name='Application Servers'),
+    ]
+    ds_db = MiddlewareDatasource.datasources_in_db(provider=provider)
+    assert len(ds_db) > 0, "There is no datasource(s) available in UI"
+    datasource = get_random_list(ds_db, 1)[0]
+    datasource.validate_tags(tags=tags)
```
Comment (psav) [inline cfme/tests/middleware/test_middleware_datasource.py]: What adds these tags to the Datastore?

Comment (hhovsepy) [inline cfme/tests/middleware/test_middleware_datasource.py]: cfme.common.__init__.Validatable.validate_tags itself cleans tags, adds provided ones and verifies whether it is added.
did I understand your question correctly?

Comment (psav) [inline cfme/tests/middleware/test_middleware_datasource.py]: OK, this method should **definitely** be renamed. It doesn't validate. It runs an entire check and obliterates tags that may already be assigned. What should happen here is that the tags are added as part of a fixture and validate here as part of the test, assuming that adding the tests is not part of the test, if it is, a separate add, and then a separate validate and delete should exist. Like a crud test. @mfalesni @jkrocil agree?

Comment (jkrocil) [inline cfme/tests/middleware/test_middleware_datasource.py]: Completely agree with @psav. Method in question: https://github.com/ManageIQ/integration_tests/blob/d21fb7af6282e7c1ba0510307020b2f4759f00c6/cfme/common/__init__.py#L384
Adding tags is one operation, removing them another and validating them another. Creating and deleting stuff is usually done in fixtures and considered test setup while validating them would be the test itself. If creating or deleting tags is the aim of the test then it should use methods that are named appropriately, not do it through a validate method.

Comment (hhovsepy) [inline cfme/tests/middleware/test_middleware_datasource.py]: I am agree as well, doing refactoring now, also moving all tag tests into single file.

**Comment by psav** [comment]
Thanks @hhovsepy

**Comment by hhovsepy** [comment]
@psav thanks for review, have refactored and squashed.
In fact 1 test fails because requires https://github.com/ManageIQ/integration_tests/pull/3180 fixes.

**Comment by dajoRH** [comment]
Lint report for commit 437755b77a14f75a019e8640ad0ee4e12b66f2b1:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 93d241cbcdcac9838fb81f6c732c2c238cf15b14:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by hhovsepy** [comment]
I can see that tag failures are because of timeout exception in code:
try:
                sel.wait_for_element(loc)
            except TypeError:
                # TypeError - when loc is not resolvable to an element, elements() will yell
                # vvv An alternate scenario when element is not resolvable, just wait a bit.
                time.sleep(1)

oc method: cfme.web_ui.__init__._fill_form_list

It tries to find "Select" field of tags and timeouts, but this field exists on opened page.

**Comment by dajoRH** [comment]
Lint report for commit b9acbafcf3f2a5ffaafd210f774140bd54822fd4:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by hhovsepy** [comment]
test_middleware_tags.py tests are passing, the failed tests are fixed in PR https://github.com/ManageIQ/integration_tests/pull/3203

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by hhovsepy** [comment]
@psav I have changed according to comments.

**Comment by dajoRH** [comment]
Lint report for commit 1db481ec756e329eb655725b0137e686206508b7:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit c1a9f4aab0d5a87fd05787038012e88e8041b94e:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8af95cac02393c1d4fe8286424da08fb28b438ae:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 80bb59232772b0b52418847322d0336f6522211a:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by hhovsepy** [comment]
@psav @mfalesni this PR become blocker for us, please take a look

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 52a743aeabaacbefb5e6507925e338d7a047cd82:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 62911d366d13007b35003355662ffe9ba9c843f2:

Everything seems all right :smile: .
*CFME QE Bot*

File: utils/providers.py [hunk 2]
```diff
@@ -368,7 +368,7 @@ def clear_provider_by_type(prov_class, validate=True):
         logger.info(' Providers exist, so removing all {} providers'.format(prov_class))
         paginator.results_per_page('100')
         sel.click(paginator.check_all())
-        toolbar.select('Configuration', 'Remove {} Providers from the VMDB'.format(string_name),
+        toolbar.select('Configuration', 'Remove {} Providers'.format(string_name),
```
Comment (psav) [inline utils/providers.py]: Are you sure this has changed in ALL version of the appliance?

Comment (hhovsepy) [inline utils/providers.py]: @psav according to this commit, it is general change for all: https://github.com/ManageIQ/manageiq/commit/4cdf4bb3fc6c53b132429f0830370c431fc631f6

Comment (psav) [inline utils/providers.py]: Yes, but this change may not have hit downstream yet, we need to ensure it's been merged :)

Comment (hhovsepy) [inline utils/providers.py]: @psav what if we check version here for upstream?

Comment (psav) [inline utils/providers.py]: yup a version pick here should be good

**Comment by dajoRH** [comment]
Lint report for commit db833cb03988a910bc5f89b1017ba549f501097e:

Everything seems all right :smile: .
*CFME QE Bot*

File: utils/providers.py [hunk 3]
```diff
@@ -368,8 +368,12 @@ def clear_provider_by_type(prov_class, validate=True):
...
         paginator.results_per_page('100')
         sel.click(paginator.check_all())
-        toolbar.select('Configuration', 'Remove {} Providers from the VMDB'.format(string_name),
-                       invokes_alert=True)
+        if version.current_version() == version.LATEST:
+            toolbar.select('Configuration', 'Remove {} Providers'.format(string_name),
+                    invokes_alert=True)
+        else:
+            toolbar.select('Configuration', 'Remove {} Providers from the VMDB'.format(string_name),
+                    invokes_alert=True)
```
Comment (mfalesni) [inline utils/providers.py]: you don't have to do an explicit version pick using if here. 
```python
toolbar.select(
    'Configuration', {
        version.LOWEST: 'Remove {} Providers from the VMDB'.format(string_name),
        version.LATEST: 'Remove {} Providers from the VMDB'.format(string_name),
    },
    invokes_alert=True)
```

Toolbar can take dictionaries and considers tehm to be verpicked.

**Comment by dajoRH** [comment]
Lint report for commit 30f039f82989a8fd51b3b380572cb0bdcf4f7d0b:

Everything seems all right :smile: .
*CFME QE Bot*

**Pull request merged**